src/client.js

```diff
@@ -0,0 +1,124 @@
+const io = require('socket.io-client')
+const { Game } = require('./Game.js')
+const C = require('./constants.js')
+
+const game = new Game({ size: 30 })
+game.turns = []
+var prevTurn = 0
+
+const socket = io()
+socket.on('game:state', (state, turnIndex) => {
+  game.players = state.players
+  game.turn = state.turn
+  if (turnIndex < game.turns.length) game.turns = []
+  game.turns[turnIndex] = state.turn
+  prevTurn = turnIndex - 1
+})
+
+const myCanvas = document.getElementById('myCanvas')
+myCanvas.width = myCanvas.offsetWidth
+myCanvas.height = myCanvas.offsetHeight
+const ctx = myCanvas.getContext('2d')
+const colors = [C.BLACK, C.RED, C.BLUE, 'cyan', 'purple', 'yellow', 'orange', 'green', 'pink', 'grey', 'teal', 'brown']
+const offset = 1
+const edge = Math.ceil((myCanvas.width / 34))
+const trailH = document.getElementById('trailHorizontal')
+const trailV = document.getElementById('trailVertical')
+const trailCornerTl = document.getElementById('trailCornerTl')
+const trailCornerTr = document.getElementById('trailCornerTr')
+const trailCornerBl = document.getElementById('trailCornerBl')
+const trailCornerBr = document.getElementById('trailCornerBr')
+const trailBuffer = {}
+
+window.requestAnimationFrame(renderGame)
+function renderGame () {
+  window.requestAnimationFrame(renderGame)
+  ctx.clearRect(0, 0, myCanvas.width, myCanvas.height)
+  const turn = game.turn
+  for (let i = 0; i < turn.board.length; ++i) {
+    const row = turn.board[i]
+    for (let j = 0; j < row.length; ++j) {
+      const cell = row[j]
+      const key = i + '_' + j
+      const bike = turn.bikes[cell - 1]
+      if (cell > C.EMPTY_CELL && turn.bikes && bike) {
```

Comment:
Checking `turn.bikes` inside the `if` is unnecessary – if it were `undefined`, it would have crashed in the previous line already (`turn.bikes[]`).